### ISSUE TITLE
Flake in guarddog_impl_test

### ISSUE THREAD

**Issue by htuch** [lyft/envoy#1141]
Triggered in https://github.com/lyft/envoy/pull/1134, see https://travis-ci.org/lyft/envoy/jobs/244689841:

[----------] 2 tests from GuardDogMissTest
[ RUN      ] GuardDogMissTest.MissTest
external/envoy/test/server/guarddog_impl_test.cc:164: Failure
      Expected: 1UL
      Which is: 1
To be equal to: stats_store_.counter("server.watchdog_miss").value()
      Which is: 2
[  FAILED  ] GuardDogMissTest.MissTest (647 ms)

**Comment by htuch** [comment]
@dnoe What do you reckon?

**Comment by dnoe** [comment]
I took a look at the code to refresh my memory.  The test uses a mock timer and over-rides the condition_variable.wait_for() timeout by early triggering the condition variable after advancing the mock timer.  But the condition variable timeout IS still active, and it's 500ms per the test.  So I suspect the scenario is that under heavy load the wait_for timeout is expiring and then the loop runs again due to the force check and it will count a miss in both circumstances because the mock time is sufficiently advanced to count a miss.  This would result in the watchdog_miss timeout incrementing twice instead of just once which is what you see in the actual vs. expected results.

This could be "fixed" by making the test allow one-or-greater (not crazy about this solution).  Or the implementation could be changed to only count one miss per forward increment of the last watchdog timestamp.   There could be over counting of misses in a real scenario (if the watched thread has not responded for many 500ms intervals, it will count each as a miss - plus there could be spurious wakeups on shorter intervals).  The fix would be to keep a local copy of watched_dogs_ vector that stores the time the watched thread last responded when counting a miss. The miss counter would only be incremented if the delta is greater than the threshold AND the time is greater than the last time we incremented the counter (ie, the watched thread responded at least one in the meantime).  I don't think the implementation is too difficult, I'll work up a PR unless you think it would be better to go for the simple "fix" to the test.
